### PR TITLE
Small Parsimony

### PR DESCRIPTION
**Pull request by mattjones315**
Implements the first section of small parsimony tools, supporting the Fitch-Hartigan algorithm and small parsimony computations.

Also adds the `tools` module which will store all post-tree-reconstruction tools (e.g., small parsimony, evolutionary coupling, etc.).

While this PR only implements the Fitch-Hartigan small-parsimony algorithm, we will continue to populate this library with several post-reconstruction tools.

### CODE REVIEW

File: cassiopeia/tools/small_parsimony.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+            of the meta data is not categorical.
+    """
+
+    if meta_item not in cassiopeia_tree.cell_meta.columns:
+        raise CassiopeiaError("Meta item does not exist in the cassiopeia tree")
+
+    meta = cassiopeia_tree.cell_meta[meta_item]
+
+    if meta.dtype not in ['str', 'object']:
+        raise CassiopeiaError("Meta item is not a categorical variable.")
```
Comment (Lioscro) [inline cassiopeia/tools/small_parsimony.py:95]: How about `category` and `int` dtypes?
Also, what's the reason for including a generic `object` dtype here?

File: cassiopeia/tools/small_parsimony.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+            categorical variable.
+        add_key: Key to add for bottom-up reconstruction
+        copy: Modify the tree in place or not.
+
+    Returns:
+        A new CassiopeiaTree if the copy is set to True, else None.
+
+    Raises:
+        CassiopeiaError if the tree does not have the specified meta data
+            of the meta data is not categorical.
```
Comment (Lioscro) [inline cassiopeia/tools/small_parsimony.py:86]: `of` -> `or`

File: cassiopeia/tools/small_parsimony.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+Amongst these tools are basic Fitch-Hartigan reconstruction, parsimony scoring,
+and the FitchCount algorithm described in Quinn, Jones et al, Science (2021).
+"""
+from cassiopeia.mixins.errors import CassiopeiaTreeError
+from typing import Optional
+
+import numpy as np
+
+from cassiopeia.data import CassiopeiaTree
+from cassiopeia.mixins import CassiopeiaError
```
Comment (Lioscro) [inline cassiopeia/tools/small_parsimony.py:14]: Could we combine the `CassiopeiaError` and `CassiopeiaTreeError` imports?

File: cassiopeia/tools/small_parsimony.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    
+    parsimony = 0
+    for (parent, child) in cassiopeia_tree.depth_first_traverse_edges(source=root):
+        
+        try:
+            if cassiopeia_tree.get_attribute(parent, label_key) != cassiopeia_tree.get_attribute(child, label_key):
+                parsimony += 1
+        except CassiopeiaTreeError:
+            raise CassiopeiaError(f"{label_key} does not exist for a node, "
+                                "try running Fitch-Hartigan.")
```
Comment (Lioscro) [inline cassiopeia/tools/small_parsimony.py:229]: Would also be useful to include a statement about `infer_ancestral_states` parameter. Something like `or provide infer_ancestral_states=True`.

File: cassiopeia/tools/small_parsimony.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    cassiopeia_tree = cassiopeia_tree.copy()
+
+    if infer_ancestral_states:
+        fitch_hartigan(cassiopeia_tree, meta_item, root, label_key=label_key)
+    
+    parsimony = 0
+    for (parent, child) in cassiopeia_tree.depth_first_traverse_edges(source=root):
+        
+        try:
+            if cassiopeia_tree.get_attribute(parent, label_key) != cassiopeia_tree.get_attribute(child, label_key):
```
Comment (Lioscro) [inline cassiopeia/tools/small_parsimony.py:225]: `black`ed?

File: test/tool_tests/small_parsimony_test.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+
+    def test_score_parsimony(self):
+
+        self.assertRaises(CassiopeiaError, cas.tl.score_small_parsimony, self.binary_tree, "nucleotide", None, False, "label")
+
+        parsimony = cas.tl.score_small_parsimony(self.binary_tree, "nucleotide", infer_ancestral_states=True)
+        self.assertEqual(parsimony, 3)
+
+        fitch_tree = cas.tl.fitch_hartigan(self.binary_tree, "nucleotide", label_key = 'inferred_nucleotide', copy=True)
+        parsimony = cas.tl.score_small_parsimony(fitch_tree, "nucleotide", infer_ancestral_states=False, label_key = "inferred_nucleotide")
```
Comment (Lioscro) [inline test/tool_tests/small_parsimony_test.py:188]: `black`ed?

**Review by Lioscro — CHANGES_REQUESTED** [review]
Excited to see this making its way into the codebase!
Just some questions and comments I'd like to get addressed.

**Comment by Lioscro** [comment]
Oh, also, I notice in the description you mention evolutionary coupling, but I don't remember seeing anything about that in the changes. Are you still trying to get those in this PR?

File: cassiopeia/tools/small_parsimony.py (hunk 1 above)
Comment (mattjones315) [reply cassiopeia/tools/small_parsimony.py:95]: You're right - I should really only be supporting `categories`.

Small-parsimony only works for categorical data, so it should break if you pass it an int or float.

File: cassiopeia/tools/small_parsimony.py (hunk 3 above)
Comment (mattjones315) [reply cassiopeia/tools/small_parsimony.py:14]: Done.

**Comment by mattjones315** [comment]
Thanks for the review Joseph! I believe I've hit all of your comments.

To note, this PR is only the first of several additions to the `tools` directory. Currently, I'm only implementing Fitch-Hartigan because I don't want to do too much in one PR. I'll move onto FitchCount and other tools in subsequent PRs.

**Review by Lioscro — APPROVED** [review]
👍

**Pull request merged**